src/lib/sb-file-uploader-hoc.jsx

```diff
@@ -0,0 +1,272 @@
+                this.fileToUpload = thisFileInput.files[0];
+
+                // If user owns the project, or user has changed the project,
+                // we must confirm with the user that they really intend to
+                // replace it. (If they don't own the project and haven't
+                // changed it, no need to confirm.)
+                let uploadAllowed = true;
+                if (userOwnsProject || (projectChanged && isShowingWithoutId)) {
+                    uploadAllowed = confirm( // eslint-disable-line no-alert
+                        intl.formatMessage(sharedMessages.replaceProjectWarning)
+                    );
+                }
+                if (uploadAllowed) {
+                    // cues step 4
+                    this.props.requestProjectUpload(loadingState);
+                } else {
+                    // skips ahead to step 7
+                    this.removeFileObjects();
+                }
+                this.props.closeFileMenu();
+            }
+        }
+        // step 4 is below, in mapDispatchToProps
+
+        // step 5: called from componentDidUpdate when project state shows
+        // that project data has finished "uploading" into the browser
+        handleFinishedLoadingUpload () {
+            if (this.fileToUpload && this.fileReader) {
+                // begin to read data from the file. When finished,
+                // cues step 6 using the reader's onload callback
+                this.fileReader.readAsArrayBuffer(this.fileToUpload);
+            } else {
+                this.props.cancelFileUpload(this.props.loadingState);
+                // skip ahead to step 7
+                this.removeFileObjects();
+            }
+        }
+        // used in step 6 below
+        getProjectTitleFromFilename (fileInputFilename) {
+            if (!fileInputFilename) return '';
+            // only parse title with valid scratch project extensions
+            // (.sb, .sb2, and .sb3)
+            const matches = fileInputFilename.match(/^(.*)\.sb[23]?$/);
+            if (!matches) return '';
+            return matches[1].substring(0, 100); // truncate project title to max 100 chars
+        }
+        // step 6: attached as a handler on our FileReader object; called when
+        // file upload raw data is available in the reader
+        onload () {
+            if (this.fileReader) {
+                this.props.onLoadingStarted();
+                const filename = this.fileToUpload && this.fileToUpload.name;
+                this.props.vm.loadProject(this.fileReader.result)
+                    .then(() => {
+                        if (filename) {
+                            const uploadedProjectTitle = this.getProjectTitleFromFilename(filename);
+                            this.props.onSetProjectTitle(uploadedProjectTitle);
+                        }
+                        this.props.onLoadingFinished(this.props.loadingState, true);
```

Comment:
I think it might be a better design to move this into the final `.then()`, since you want to ensure that either it will be called in the `catch` OR it will be called when it's determined the promise chain was successful. Then it's safer for that pattern to be repeated in other places if it gets copied around and the details of the error-throwing code changes.

Reply:
That makes sense, thanks!